src/MixedMatrix.cpp

```diff
@@ -15,221 +15,311 @@
 
 /** @file
 
-    @brief Implements MixedMatrix object.
+    @brief MixedMatrix class
 */
 
-#ifndef __MIXEDMATRIX_IMPL_HPP__
-#define __MIXEDMATRIX_IMPL_HPP__
-
 #include "MixedMatrix.hpp"
-#include "mfem.hpp"
-#include "utilities.hpp"
-#include <memory>
-
-using std::unique_ptr;
 
 namespace smoothg
 {
 
-MixedMatrix::MixedMatrix(const mfem::SparseMatrix& vertex_edge,
-                         const mfem::Vector& weight,
-                         const mfem::SparseMatrix& w_block,
-                         const mfem::HypreParMatrix& edge_d_td,
-                         DistributeWeight dist_weight)
-    : edge_d_td_(&edge_d_td),
-      edge_td_d_(edge_d_td_->Transpose())
+MixedMatrix::MixedMatrix(const Graph& graph)
+    : edge_true_edge_(graph.edge_true_edge_),
+      D_local_(MakeLocalD(graph.edge_true_edge_, graph.vertex_edge_local_)),
+      W_local_(graph.W_local_),
+      elem_dof_(graph.vertex_edge_local_),
+      agg_vertexdof_(SparseIdentity(D_local_.Rows())),
+      face_facedof_(SparseIdentity(elem_dof_.Cols()))
```

Comment:
I guess `agg_vertexdof_` and `face_facedof_` come from the old code (maybe `HybridSolver`). They are pretty bad names (I think I am the one who gave them these names) in the sense that there is no `agg` nor `face` in a certain level. They are `vertex_vdof` `edge_edof` in the multilevel code.

`elem_dof` is a bit trickier, it should be called `vertex_edof` if we follow the same convention above. But this will have a naming conflict with the ml code (`vertex_edof` means something else in ml code), so lets leave it as it is for now. But I think we need to do something to it later.

Reply:
Yes, there are a bunch more relationships from the ML code that can be aggregated into one structure later on.  You had `GraphSpace` before, or something like that.